src/test/shell/bazel/bazel_symlink_test.sh

```diff
@@ -716,4 +721,179 @@ EOF
   bazel build --experimental_allow_unresolved_symlinks //a:exec || fail "build failed"
 }
 
+function setup_unresolved_symlink_as_input() {
+  add_to_bazelrc build --experimental_allow_unresolved_symlinks
+
+  mkdir -p pkg
+  cat > pkg/def.bzl <<'EOF'
+def _r(ctx):
+  symlink = ctx.actions.declare_symlink(ctx.label.name + "_s")
+  ctx.actions.symlink(output = symlink, target_path = ctx.file.file.basename)
+
+  output = ctx.actions.declare_file(ctx.label.name)
+  ctx.actions.run_shell(
+    command = """[ -s {symlink} ] && [ $(readlink {symlink}) == "{target}" ] && touch {output}""".format(
+      symlink = symlink.path,
+      target = ctx.file.file.basename,
+      output = output.path,
+    ),
+    inputs = [symlink] + ([ctx.file.file] if ctx.attr.stage_target else []),
+    outputs = [output],
+  )
+  return DefaultInfo(files=depset([output]))
+
+r = rule(
+    implementation = _r,
+    attrs = {
+        "file": attr.label(allow_single_file=True),
+        "stage_target": attr.bool(),
+    }
+)
+EOF
+  cat > pkg/BUILD <<'EOF'
+load(":def.bzl", "r")
+
+genrule(name="a", outs=["file"], cmd="echo hello >$@")
+r(name="b", file="file", stage_target=False)
+r(name="c", file="file", stage_target=True)
+EOF
+
+}
+
+function test_unresolved_symlink_as_input_local() {
+  if "$is_windows"; then
+    # TODO(#10298): Support unresolved symlinks on Windows.
+    return 0
+  fi
+
+  setup_unresolved_symlink_as_input
+  add_to_bazelrc build --spawn_strategy=local
+
+  bazel build //pkg:b && fail "symlink should not resolve"
+
+  bazel clean
+  bazel build //pkg:file
+  ls -R bazel-out
+  # Since the build isn't sandboxed, the symlink to //:a resolves even though
+  # the action does not declare it as an input.
+  bazel build //pkg:b || fail "symlink expected to resolve non-hermetically"
+
+  bazel clean
+  bazel build //pkg:c || fail "symlink should resolve"
+}
+
+function test_unresolved_symlink_as_input_sandbox() {
+  if "$is_windows"; then
+    # TODO(#10298): Support unresolved symlinks on Windows.
+    return 0
+  fi
+
+  setup_unresolved_symlink_as_input
+  add_to_bazelrc build --spawn_strategy=sandboxed
+
+  bazel build //pkg:b && fail "sandboxed build is not hermetic"
+
+  bazel clean
+  bazel build //pkg:a
+  # Since the build isn't sandboxed, the symlink to //:a does not resolves even
+  # though it would in the unsandboxed exec root due to //:a having been built
+  # before.
+  bazel build //pkg:b && fail "sandboxed build is not hermetic"
+
+  bazel clean
+  bazel build //pkg:c || fail "symlink should resolve"
+}
+
+function setup_unresolved_symlink_as_runfile() {
+  add_to_bazelrc build --experimental_allow_unresolved_symlinks
+
+  mkdir -p pkg
+  cat > pkg/def.bzl <<'EOF'
+def _r(ctx):
+  symlink = ctx.actions.declare_symlink(ctx.label.name + "_s")
+  target = ctx.file.file.basename
+  ctx.actions.symlink(output=symlink, target_path=target)
+
+  script = ctx.actions.declare_file(ctx.label.name)
+  content = """
+#!/usr/bin/env bash
```

Comment:
Mind extracting this to an input file to avoid quoting shell quoting Starlark quoting shell again?

Reply:
I extracted it into a template for `ctx.actions.expand_template`. Is that what you meant?

Reply:
I was confused for a moment when you mentioned `expand_template` but after looking the code, it makes total sense. I'd probably have used command line arguments instead of template expansion to inject the paths but that's probably just a matter of taste and the main issue of triple-quoting goes away in either case.